### PR TITLE
Add type-inference tests

### LINKED ISSUE #1021: Also run the type-inference tests
I noticed that `dhall:test:tasty` doesn't run these: https://github.com/dhall-lang/dhall-lang/tree/master/tests/type-inference. But maybe there's a reason?!

### PR DESCRIPTION
**Pull request by sjakobi**
Fixes #1021.

### CODE REVIEW

File: dhall/tests/Dhall/Test/TypeInference.hs [hunk 1]
```diff
@@ -0,0 +1,50 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Dhall.Test.TypeInference where
```
Comment (sjakobi) [inline dhall/tests/Dhall/Test/TypeInference.hs]: This is simply copied from `Dhall.Test.TypeCheck`.

**Comment by sjakobi** [comment]
There are a few failures, all with `Error: ❰Sort❱ has no type, kind, or sort`:

```
Dhall Tests
  type-inference tests
    discover
      ./dhall-lang/tests/type-inference/success/unit/RecordTypeKindLike:                        FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/RecordTypeKindLikeB.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1: Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/RecordTypeKindLikeB.dhall:1:1
      ./dhall-lang/tests/type-inference/success/unit/RecordTypeNestedKind:                      FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/RecordTypeNestedKindB.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1: Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/RecordTypeNestedKindB.dhall:1:1
      ./dhall-lang/tests/type-inference/success/unit/FunctionTypeKindKind:                      FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/FunctionTypeKindKindB.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1: Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/FunctionTypeKindKindB.dhall:1:1
      ./dhall-lang/tests/type-inference/success/unit/FunctionTypeKindType:                      FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/FunctionTypeKindTypeB.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1: Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/FunctionTypeKindTypeB.dhall:1:1
      ./dhall-lang/tests/type-inference/success/unit/Kind:                                      FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/KindB.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1: Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/KindB.dhall:1:1
      ./dhall-lang/tests/type-inference/success/unit/RecordTypeKind:                            FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/RecordTypeKindB.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1: Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/RecordTypeKindB.dhall:1:1
      ./dhall-lang/tests/type-inference/success/unit/RecursiveRecordTypeMergeTwoKinds:          FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/RecursiveRecordTypeMergeTwoKindsB.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1: Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/RecursiveRecordTypeMergeTwoKindsB.dhall:1:1
      ./dhall-lang/tests/type-inference/success/unit/UnionTypeKind:                             FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/UnionTypeKindB.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1: Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/UnionTypeKindB.dhall:1:1
      ./dhall-lang/tests/type-inference/success/unit/RecordTypeNestedKindLike:                  FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/RecordTypeNestedKindLikeB.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1: Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/RecordTypeNestedKindLikeB.dhall:1:1
      ./dhall-lang/tests/type-inference/success/unit/TypeAnnotationSort:                        FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/TypeAnnotationSortA.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1:        Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/TypeAnnotationSortA.dhall:1:8
      ./dhall-lang/tests/type-inference/success/unit/RecursiveRecordTypeMergeRecursivelyKinds:  FAIL
        Exception: 
        ↳ ./dhall-lang/tests/type-inference/success/unit/RecursiveRecordTypeMergeRecursivelyKindsB.dhall
        
        Error: ❰Sort❱ has no type, kind, or sort
        
        1: Sort
        
        /home/simon/src/dhall-haskell/dhall/dhall-lang/tests/type-inference/success/unit/RecursiveRecordTypeMergeRecursivelyKindsB.dhall:1:1

11 out of 1000 tests failed (3.12s)
```

File: dhall/tests/Dhall/Test/TypeInference.hs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    successTests <- Test.Util.discover (Turtle.chars <* "A.dhall") successTest (Turtle.lstree (typecheckDirectory </> "success"))
+
+    let testTree = Tasty.testGroup "type-inference tests"
+            [ successTests
+            ]
+
+    return testTree
+
+successTest :: Text -> TestTree
+successTest prefix =
```
Comment (sjakobi) [inline dhall/tests/Dhall/Test/TypeInference.hs]: This doesn't currently run the right checks:

```
### Running `type-inference` tests

The tests should:
- parse `A` and `B`
- infer the type of A
- the inferred type of A should exactly match B
```

**Comment by sjakobi** [comment]
Only one failure remains:

```
Dhall Tests
  type-inference tests
    discover
      ./dhall-lang/tests/type-inference/success/unit/TypeAnnotationSort:                        FAIL
        Exception: 
        Error: ❰Sort❱ has no type, kind, or sort
```

**Comment by Gabriel439** [comment]
@sjakobi: That can be fixed by implementing https://github.com/dhall-lang/dhall-lang/pull/507

Basically, there just needs to be a special type-checking rule added for the case of a type annotation of `Sort`

**Comment by sjakobi** [comment]
> That can be fixed by implementing [dhall-lang/dhall-lang#507](https://github.com/dhall-lang/dhall-lang/pull/507)

Cool! I've made a PR: https://github.com/dhall-lang/dhall-haskell/pull/1024

**Comment by Gabriel439** [comment]
@sjakobi: Also, you will need to add the files underneath `./dhall-lang/tests/type-inference/success` to the `extra-source-files` section of the `dhall.cabal` file

**Comment by sjakobi** [comment]
> @sjakobi: Also, you will need to add the files underneath `./dhall-lang/tests/type-inference/success` to the `extra-source-files` section of the `dhall.cabal` file

Thanks. I had been wondering why `cabal` was complaining about missing files!

I still can't run the doctests with `cabal` BTW:

```
$ cabal test dhall:test:doctest
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - dhall-1.24.0 (test:doctest) (first run)
./dhall.cabal has been changed. Re-configuring with most recently used
options. If this fails, please run configure manually.
Configuring test suite 'doctest' for dhall-1.24.0..
Preprocessing test suite 'doctest' for dhall-1.24.0..
Building test suite 'doctest' for dhall-1.24.0..
Running 1 test suites...
Test suite doctest: RUNNING...

/home/simon/src/dhall-haskell/dhall/src/Dhall/Parser.hs:47:0: error:
     error: missing binary operator before token "("
     #if MIN_VERSION_megaparsec(7, 0, 0)
     ^
   |
47 | #if MIN_VERSION_megaparsec(7, 0, 0)
   | ^

/home/simon/src/dhall-haskell/dhall/src/Dhall/Parser.hs:57:0: error:
     error: missing binary operator before token "("
     #if MIN_VERSION_megaparsec(7, 0, 0)
     ^
   |
57 | #if MIN_VERSION_megaparsec(7, 0, 0)
   | ^
doctest: `gcc' failed in phase `C pre-processor'. (Exit code: 1)
Test suite doctest: FAIL
Test suite logged to:
/home/simon/src/dhall-haskell/dist-newstyle/build/x86_64-linux/ghc-8.6.5/dhall-1.24.0/t/doctest/test/dhall-1.24.0-doctest.log
0 of 1 test suites (0 of 1 test cases) passed.
cabal: Tests failed for test:doctest from dhall-1.24.0.
```

stack works though!

**Comment by Gabriel439** [comment]
@sjakobi: What version of `doctest` are you using?

**Comment by sjakobi** [comment]
> What version of `doctest` are you using?

`v0.16.1` with `cabal`. It's the same with `v.16.0.1` which the `stack.yaml` defines though.

**Comment by sjakobi** [comment]
> By any chance are you using Cabal's v2 (a.k.a. "new-build") commands?

Yep. And I believe I've seen that doctest issue in other projects before… stack to the rescue! :)

**Comment by Gabriel439** [comment]
Yeah, then that seems to be the common thread here: `stack` and Nix use Cabal's v1 commands whereas you and @blamario have been using the v2 commands

**Pull request merged**